src/main/java/com/epam/reportportal/karate/ReportPortalPublisher.java

```diff
@@ -0,0 +1,257 @@
+import com.epam.ta.reportportal.ws.model.launch.StartLaunchRQ;
+import com.epam.ta.reportportal.ws.model.log.SaveLogRQ;
+import com.intuit.karate.core.*;
+import io.reactivex.Maybe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class ReportPortalPublisher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReportPortalPublisher.class);
+    private static final String INFO_LOG_LEVEL = "INFO";
+    private static final String ERROR_LOG_LEVEL = "ERROR";
+    private static final String WARN_LOG_LEVEL = "WARN";
+    private static final String PASSED = "PASSED";
+    private static final String FAILED = "FAILED";
+    private static final String SKIPPED = "SKIPPED";
+    private static final String STOPPED = "STOPPED";
+    private static final String RESETED = "RESETED";
+    private static final String CANCELLED = "CANCELLED";
+
+    private final Supplier<Launch> launch;
+    private final ConcurrentHashMap<String, Maybe<String>> featureIdMap;
+    private final ConcurrentHashMap<String, Maybe<String>> scenarioIdMap;
+    private final ConcurrentHashMap<Maybe<String>, Long> stepStartTimeMap;
+    private Maybe<String> stepId;
+
+    public ReportPortalPublisher() {
+        this(createLaunchSupplier());
+    }
+
+    public ReportPortalPublisher(Supplier<Launch> launch) {
+        this.launch = launch;
+        this.featureIdMap = new ConcurrentHashMap<>();
+        this.scenarioIdMap = new ConcurrentHashMap<>();
+        this.stepStartTimeMap = new ConcurrentHashMap<>();
+    }
+
+    private static Supplier<Launch> createLaunchSupplier() {
+        return new Supplier<>() {
+            private final Launch launch = createLaunch();
+
+            @Override
+            public Launch get() {
+                return launch;
+            }
+        };
+    }
+
+    private static Launch createLaunch() {
+        final ReportPortal reportPortal = ReportPortal.builder().build();
+        ListenerParameters parameters = reportPortal.getParameters();
+        StartLaunchRQ rq = new StartLaunchRQ();
+        rq.setName(parameters.getLaunchName());
+        rq.setStartTime(Calendar.getInstance().getTime());
+        rq.setMode(parameters.getLaunchRunningMode());
+        rq.setAttributes(new HashSet<>(parameters.getAttributes()));
+        if (!isNullOrEmpty(parameters.getDescription())) {
+            rq.setDescription(parameters.getDescription());
+        }
+        rq.setStartTime(Calendar.getInstance().getTime());
+        rq.setRerun(parameters.isRerun());
+        if (isNotBlank(parameters.getRerunOf())) {
+            rq.setRerunOf(parameters.getRerunOf());
+        }
```

Comment:
Here and for the following code please move all request construction logic into separate corresponding protected methods. As here: https://github.com/reportportal/agent-java-testNG/blob/develop/src/main/java/com/epam/reportportal/testng/TestNGService.java#L303

Or here: https://github.com/reportportal/agent-java-junit5/blob/develop/src/main/java/com/epam/reportportal/junit5/ReportPortalExtension.java#L125

This is necessary for class extension.

Reply:
Refactored. Fixed.